### PR TITLE
Try to minimize the external dependencies for OMPI wrappers.

### PR DESCRIPTION
**Pull request by bosilca**
The solution put forward here consist of creating a convenience library,
opal-tiny, that holds all the basic things we need to have a working
app with minimal external dependencies.

More detailed discussion in #9869.

Signed-off-by: George Bosilca <bosilca@icl.utk.edu>

### CODE REVIEW

**Comment by bosilca** [comment]
There is one point that needs to be clarified: why we need to register our output files with PMIX.

File: opal/util/proc.c [hunk 1]
```diff
@@ -215,14 +198,36 @@ char *opal_get_proc_hostname(const opal_proc_t *proc)
     if (proc == opal_proc_my_name) {
         return strdup(opal_process_info.nodename);
     }
+    return strdup("unknown");
+#if 0
```
Comment (bwbarrett) [inline opal/util/proc.c:202]: We shouldn't #if 0 code, but just remove it.  But we should really add a comment about what this function does in the new scope (and that it's replaced when using the full library).  But bigger picture, why is this code even in the util library?

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/da05bde09b142e948898e3900c5ef795

File: opal/util/output.c [hunk 2]
```diff
@@ -782,9 +782,13 @@ static int open_file(int i)
             free(filename); /* release the filename in all cases */
             return OPAL_ERR_IN_ERRNO;
         }
-
+#if 0
```
Comment (bwbarrett) [inline opal/util/output.c:785]: This prevents PMIx from deleting these files during cleanup (since they all end up in the PMIx directory in many situations).  We don't want to lose this functionality, but it makes the util / non-util OPAL split harder.  I think the right answer is yet another function pointer.  have an opal_pmix_opal_output_handler() function in pmix base that is set as the function pointer in the util/output.c code when the opal_init() call occurs (as opposed to the opal_init_util() cal).  We don't need this for utilities that aren't going to use PMIx, agree there.

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/e5e90c92a9e9e7e1e307bd6c5cb5d223

File: opal/tools/wrappers/Makefile.am [hunk 3]
```diff
@@ -59,7 +59,7 @@ endif # OPAL_INSTALL_BINARIES
 endif # OPAL_WANT_SCRIPT_WRAPPER_COMPILERS
 
 opal_wrapper_SOURCES = opal_wrapper.c
-opal_wrapper_LDADD = $(top_builddir)/opal/lib@OPAL_LIB_NAME@.la
+opal_wrapper_LDADD = $(top_builddir)/opal/lib@OPAL_LIB_NAME@_tiny.la
```
Comment (bwbarrett) [inline opal/tools/wrappers/Makefile.am:62]: I'd prefer `_util`, to match the initialization bits.

File: opal/util/proc.h [hunk 4]
```diff
@@ -171,6 +171,6 @@ OPAL_DECLSPEC extern struct opal_proc_t *(*opal_proc_for_name)(const opal_proces
  * our own. This is to be used by all BTLs so we don't retrieve hostnames
  * unless needed. The returned value MUST NOT be free'd as it is
  * owned by the proc_t */
-OPAL_DECLSPEC char *opal_get_proc_hostname(const opal_proc_t *proc);
+OPAL_DECLSPEC extern char *(*opal_get_proc_hostname)(const opal_proc_t *proc);
```
Comment (bwbarrett) [inline opal/util/proc.h:174]: it feels like the proc code shouldn't be in the util library.  I'm guessing there's a dependency I'm missing, but that would certainly be my preference on a split point.

File: opal/Makefile.am [hunk 5]
```diff
@@ -70,30 +70,49 @@ DIST_SUBDIRS = \
...
+        mca/timer/libmca_timer.la \
+        $(MCA_opal_timer_STATIC_LTLIBS) \
+        mca/installdirs/libmca_installdirs.la \
+        $(MCA_opal_installdirs_STATIC_LTLIBS) \
+        mca/if/libmca_if.la \
+        $(MCA_opal_if_STATIC_LTLIBS) \
+        mca/dl/libmca_dl.la \
+        $(MCA_opal_dl_STATIC_LTLIBS) \
+        $(opal_libevent_LIBS)
+lib@OPAL_LIB_NAME@_tiny_la_DEPENDENCIES = \
```
Comment (bwbarrett) [inline opal/Makefile.am:92]: we shouldn't need to declare dependencies, and definitely not this subset.

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/63423e6ee0481c5803b7bc1aaab2e605

File: opal/mca/pmix/base/pmix_base_fns.c [hunk 6]
```diff
@@ -542,3 +542,4 @@ static void infoitdecon(opal_info_item_t *p)
 OBJ_CLASS_INSTANCE(opal_info_item_t, opal_list_item_t, infoitmcon, infoitdecon);
 
 OBJ_CLASS_INSTANCE(opal_proclist_t, opal_list_item_t, NULL, NULL);
+
```
Comment (bwbarrett) [inline opal/mca/pmix/base/pmix_base_fns.c:545]: unnecessary change?

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: opal/util/output.c (hunk 2 above)
Comment (bosilca) [reply opal/util/output.c:785]: I'm afraid that is not enough, for any software that explicitly calls `opal_util_init` before `opal_init`. We could build a temporary list and once the RTE is initialized migrate all files there, but this seems extremely invasive.

File: opal/util/proc.h (hunk 4 above)
Comment (bosilca) [reply opal/util/proc.h:174]: I tried that, but there is no easy way because util being used via SUBDIRS, it generates a convenience library. If we want to split this, we will have to generate 2 convenience libraries, and include them accordingly at the upper level.

File: opal/util/output.c (hunk 2 above)
Comment (bwbarrett) [reply opal/util/output.c:785]: what if we didn't have libopen_pal.la include libopen_pal_util.la?  We can have libopen_pal.la include all the same source files as libopenpal_util.la, except for a single file which provided opal_pmix_register_cleanup() as an empty function?  A better solution would be weak symbols, but since Darwin doesn't support weak symbols, that's out.

File: opal/util/proc.h (hunk 4 above)
Comment (bwbarrett) [reply opal/util/proc.h:174]: I was originally leaning that way (splitting the util directory).  I'm out of time for a bit, but tonight I'll try to see if we can get rid of the subdirs for the util directory.  It's either because of the lex file (harder to get rid of) or the LTDLINCL, which shouldn't be needed at this point.

**Comment by gpaulsen** [comment]
I linked this to issue #9869.  Is that correct?

**Comment by wckzhang** [comment]
What's the status on this PR, are we agreed that this is the correct fix for #9869 ?

**Comment by bwbarrett** [comment]
I started https://github.com/bwbarrett/ompi/tree/bugfix/mpicc-dependencies to make this a bit more clean to merge into main, but have run out of time and I'm not sure when I'll be able to come back to it.  I think most of the hard problems are solved in that branch, but there's definitely some cleanup work in all the patches before it can become a PR.
